### ISSUE TITLE
[$250] Chat- "New messages" doesn't jump directly to the message marked as Unread for second time

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#12249]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Issue found when executing PR** https://github.com/Expensify/App/pull/12124 

## Action Performed:
1. Go to staging.new.expensify.com
2. Go to any chat
3. Click at Mark as Unread
4. Click on New Messages button
5. Clicking "new messages" should jump directly to the message marked as Unread
6. Scroll to the top of the conversation
7. Click on New Messages button

## Expected Result:
"New messages" should jump directly to the message marked as Unread for second time.

## Actual Result:
"New messages" doesn't jump directly to the message marked as Unread for second time

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- Mobile Web

**Version Number:** 1.2.21.4

**Reproducible in staging?:** Yes

**Reproducible in production?:** No

**Email or phone of affected tester (no customers):** 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/198686591-b6dd47b2-4547-4ed0-89af-e9f28f24dc37.mp4

https://user-images.githubusercontent.com/93399543/198686593-c5429b38-42a8-4293-96b1-12b2255b9dca.mp4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Applause - Internal Team

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01fec5d287f2840f74</li>
        <li>Upwork Job ID: 1806803838290439356</li>
        <li>Last Price Increase: 2026-09-02</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @emkhalid</details>

**Comment by Julesssss** [comment]
I'm unable to reproduce this currently, on staging Chrome. However, I think we should keep this open until we can figure out the exact reproducable steps -- but I don't think this should be marked as a regression.

CC @srikarparsi @eVoloshchak -- it seems related to [this PR](https://github.com/Expensify/App/pull/12124), any idea how to resolve this? 

https://user-images.githubusercontent.com/10736861/199002701-41b31fa0-a443-4de0-9410-613a9814b264.mov

**Comment by eVoloshchak** [comment]
Unable to reproduce this building from the latest sources

https://user-images.githubusercontent.com/9059945/199005065-36878344-0f85-45ea-a224-657715326eaa.mov

On the first video from the description looks like the issue is with the list not being loaded completely, you can see the progress bar resizing in the middle of the animation. Not sure about the second video though

**Comment by johncschuster** [comment]
Bumping to keep Melvin happy. 

@srikarparsi, do you have any ideas on this one?

**Comment by srikarparsi** [comment]
I'm also unable to reproduce the issue in staging. @Julesssss Maybe we could close the issue and anyone can reopen it if they're able to reproduce it? 

@kbecciv can you check if this is still reproducible?

**Comment by kbecciv** [comment]
Issue is not reproductible in Web with build 1.2.23.9

https://user-images.githubusercontent.com/93399543/200014749-99c44402-065c-4a4e-8890-c86f57d0b5ac.mp4

**Comment by srikarparsi** [comment]
looks good, thanks @kbecciv. @johncschuster I think we can close this out for now and anyone can reopen if they are able to reproduce.

**Comment by kbecciv** [comment]
QA team is able to reproduce this issue in Android native on build 1.3.18.1

https://github.com/Expensify/App/assets/93399543/6b720ccb-252a-4e71-bb00-081df506931d

**Comment by johncschuster** [comment]
This is on hold, Melvin. Chillax dude.

**Comment by johncschuster** [comment]
I'm downgrading the label to `Weekly`

**Comment by johncschuster** [comment]
Stil on hold.

**Comment by johncschuster** [comment]
This is still on hold.

**Comment by johncschuster** [comment]
No update

**Comment by johncschuster** [comment]
No update

**Comment by johncschuster** [comment]
@MonilBhavsar coming from the Slack thread above ([link for ease](https://expensify.slack.com/archives/C03U7DCU4/p1685448708169149)), it sounds like there was an issue you were working on that may have fixed the behavior reported here. Can you take a look and let me know if we're good to close this issue out? Thank you!

**Comment by johncschuster** [comment]
@MonilBhavsar bump!

**Comment by MonilBhavsar** [comment]
Sorry, I missed the previous ping. Yes, please put this on HOLD, and feel free to bump it to Monthly

**Comment by johncschuster** [comment]
Sounds great! Thanks @MonilBhavsar!

**Comment by johncschuster** [comment]
Ugh... I don't recall adding `Daily` to this. Removing it.

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold. Also, I've filed this under #vip-vsb since it's related to chat.

**Comment by johncschuster** [comment]
Still on hold

**Comment by MonilBhavsar** [comment]
The parent is closed. We can retest this and most probably should be fixed

**Comment by johncschuster** [comment]
[Asking for a retest](https://expensify.slack.com/archives/C9YU7BX5M/p1716923438654829)

**Comment by kavimuru** [comment]
@johncschuster Issue doesn't seem to be fixed. Clicking on `new message` pill taking to the end of conversation and `new message` pill disappear on the first click.

https://github.com/Expensify/App/assets/43996225/7f451749-4e51-4736-9cb2-c5bed22032da

https://github.com/Expensify/App/assets/43996225/76f8c947-7bd3-4c25-ac58-be7374061468

**Comment by johncschuster** [comment]
I'm sorry I missed [this comment](https://github.com/Expensify/App/issues/12249#issuecomment-2136042992). I've just triaged this one out.

**Comment by yuriiz** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Clicking "New Messages" always scrolls to bottom while it should scroll not further than the first unread message even if it is not at the bottom.

### What is the root cause of that problem?
ReportActionList.scrollToBottomAndMarkReportAsRead() handling "New Messages" click event unconditionally calls reportScrollManager.scrollToBottom().
https://github.com/Expensify/App/blob/main/src/pages/home/report/ReportActionsList.tsx#L412

### What changes do you think we should make in order to solve the problem?
Change ReportActionList.scrollToBottomAndMarkReportAsRead() to call reportScrollManager.scrollToIndex() and pass index of the first unread item.

### What alternative solutions did you explore? (Optional)

**Comment by johncschuster** [comment]
@Ollyws can you take a look at the above proposals when you get a moment?

**Comment by johncschuster** [comment]
Not overdue! I've just put this on hold to see if https://github.com/Expensify/App/issues/35011 resolves it.

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
Still on hold for #35011. Looks like we've got some draft PRs up!

**Comment by johncschuster** [comment]
Looks like the PR in the linked issue has an edge case that's being worked through.

**Comment by johncschuster** [comment]
@janicduplessis has a PR up in draft mode.

**Comment by johncschuster** [comment]
Looks like we're waiting on a re-review of @janicduplessis's PR in the linked issue. Once that issue is resolved, let's reassess this one!

**Comment by johncschuster** [comment]
No update. We're still waiting on a re-review of the linked PR

**Comment by johncschuster** [comment]
Same update as last week

**Comment by johncschuster** [comment]
Heads up, I will be OOO from Monday, December 23, and returning Monday, January 6. I am not reassigning my assigned issues. If you feel this issue needs to be actioned by a Bug Zero team member, please reach out for help from any available team members in #expensify-open-source.

If not, I'll action this as soon as I'm back in office!

**Comment by johncschuster** [comment]
We're waiting on conflicts to be resolved in the linked issue above.

**Comment by johncschuster** [comment]
Still waiting on the conflicts to be resolved. I've bumped that issue.

**Comment by johncschuster** [comment]
There was a regression in the linked issue. A PR to fix that regression will be going out soon

**Comment by johncschuster** [comment]
Still on hold for the linked issue above. Looks like we're getting close!

**Comment by johncschuster** [comment]
Still on hold for https://github.com/Expensify/App/issues/35011 (bringing this link down so we don't need to load 42 comments)

**Comment by johncschuster** [comment]
Looks like the issue above is almost merge-ready. 🎉

**Comment by johncschuster** [comment]
Discussion is ongoing in the linked PR above.

**Comment by johncschuster** [comment]
Same as last week.

**Comment by johncschuster** [comment]
No update

**Comment by johncschuster** [comment]
Still on hold

**Comment by johncschuster** [comment]
No update on the linked issue. This is still on hold for that.

**Comment by johncschuster** [comment]
Still on hold for https://github.com/Expensify/App/issues/35011

**Comment by johncschuster** [comment]
Looks like the PRs on the linked issue are on staging. Hopefully we can take this off hold soon.

**Comment by johncschuster** [comment]
Found another issue in the linked PR above. Hopefully that PR will be deployed Monday.

**Comment by johncschuster** [comment]
Looks like we're waiting on one more fix in the linked issue above. I'll check back in to that next week.

**Comment by johncschuster** [comment]
Looks like the main PR in the linked GH is being tested. Still on hold for that!

**Comment by johncschuster** [comment]
Looks like the linked issue/PR is being reviewed. Still on hold while we wait for that.

**Comment by johncschuster** [comment]
Same as before. We're still waiting on the above linked issue to get resolved.

**Comment by johncschuster** [comment]
No update. Still waiting on the upstream issue.

**Comment by johncschuster** [comment]
No update. Still on hold.

**Comment by johncschuster** [comment]
Still on hold for https://github.com/Expensify/App/issues/35011

**Comment by johncschuster** [comment]
No update. Still on hold.

**Comment by yusufdeveloper2903** [comment]
.

**Comment by yuvrajangadsingh** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

This is the bug found while testing PR #12124: clicking "New Messages" a second time (after scrolling back to the top) does not jump to the message that was marked as Unread, because the handler always scrolls to the bottom of the conversation instead of to the unread marker's actual position.

### What is the root cause of that problem?

`scrollToBottomAndMarkReportAsRead` at `src/pages/inbox/report/ReportActionsList.tsx:660-674` unconditionally calls `reportScrollManager.scrollToBottom()` in both branches, regardless of whether an unread marker exists at a specific index:

```ts
const scrollToBottomAndMarkReportAsRead = useCallback(() => {
    setIsFloatingMessageCounterVisible(false);
    if (!hasNewestReportAction) {
        ...
        reportScrollManager.scrollToBottom(); // always newest
        return;
    }
    reportScrollManager.scrollToBottom(); // always newest, ignores unreadMarkerReportActionIndex
    readActionSkipped.current = false;
    readNewestAction(report.reportID, !!reportLoadingState?.hasOnceLoadedReportActions);
}, [...]);
```

`unreadMarkerReportActionIndex` is already computed at line 321 (via `getUnreadMarkerReportAction`) and passed into `useReportUnreadMessageScrollTracking` at line 388, but it is never used by the scroll handler. When the button shows "New Messages" (i.e., `unreadMarkerReportActionID` is set), the handler should scroll to that specific index, not to index 0.

The first click can appear to work because the marked message is close enough to the newest messages that `scrollToBottom()` lands it within the viewport. The second click fails more visibly after the user has scrolled all the way to the top, because `scrollToBottom()` goes to the newest messages (far from the marked message), and if the unread marker is more than a screenful above the bottom, it never enters view.

Note: `readNewestAction` is called without `shouldResetUnreadMarker = true` in this handler, so `unreadMarkerTime` state is never updated by this click path. The Onyx `report.lastReadTime` is updated optimistically, but `unreadMarkerReportActionID` stays non-null because `unreadMarkerTime` only resets via the `readNewestAction_${reportID}` DeviceEventEmitter event (which requires `shouldResetUnreadMarker = true`) or a report switch. This means the marker correctly persists as an anchor for repeated "jump to unread" clicks, and the button should use that anchor.

### What changes do you think we should make in order to solve the problem?

Change the `scrollToBottomAndMarkReportAsRead` callback in `src/pages/inbox/report/ReportActionsList.tsx:660-674` to scroll to the unread marker index when one exists, and fall back to `scrollToBottom()` when there is none:

```ts
const scrollToBottomAndMarkReportAsRead = useCallback(() => {
    setIsFloatingMessageCounterVisible(false);

    if (!hasNewestReportAction) {
        if (!Navigation.getReportRHPActiveRoute()) {
            Navigation.navigate(ROUTES.REPORT_WITH_ID.getRoute(report.reportID, undefined, undefined, backTo));
        }
        openReport({reportID: report.reportID, introSelected, betas});
        reportScrollManager.scrollToBottom();
        return;
    }

    if (unreadMarkerReportActionIndex !== -1) {
        reportScrollManager.scrollToIndex(unreadMarkerReportActionIndex);
    } else {
        reportScrollManager.scrollToBottom();
    }
    readActionSkipped.current = false;
    readNewestAction(report.reportID, !!reportLoadingState?.hasOnceLoadedReportActions);
}, [setIsFloatingMessageCounterVisible, hasNewestReportAction, unreadMarkerReportActionIndex, reportScrollManager, report.reportID, backTo, introSelected, reportLoadingState?.hasOnceLoadedReportActions, betas]);
```

The only addition is the `if (unreadMarkerReportActionIndex !== -1)` guard, using `scrollToIndex` which is already called at line 583 in `scrollToBottomForCurrentUserAction`. `unreadMarkerReportActionIndex` needs to be added to the `useCallback` deps. No other files need to change.

After `scrollToIndex(N)`, `onViewableItemsChanged` fires when the marked message enters the viewport, which sets `isFloatingMessageCounterVisible(false)` (button hides). If the user scrolls away again, the button reappears and clicking it jumps back to the same message. This is the correct "return to unread anchor" behavior.

### What alternative solutions did you explore? (Optional)

I considered passing `shouldResetUnreadMarker = true` to `readNewestAction` inside this handler to explicitly clear `unreadMarkerTime` after clicking, but that would cause the button to disappear permanently after one click even when the user hasn't actually read the message yet. The existing approach of letting `onViewableItemsChanged` clear the button when the marker comes into view is the right design. I also considered hooking this into `useReportUnreadMessageScrollTracking` to make the scroll-to-marker behavior part of the tracking hook, but the index is already available at the call site and the change is simpler as a direct guard.

**Comment by yuvrajangadsingh** [comment]
sorry, jumped the gun here. didn't catch that this is still on hold for #35011 (last "still on hold" from @johncschuster was Apr 22, and the linked PRs aren't confirmed deployed yet), and @yuriiz already covered the same `scrollToBottomAndMarkReportAsRead` root cause + `scrollToIndex` fix back in 2024.

withdrawing this proposal.

**Comment by johncschuster** [comment]
@yuvrajangadsingh 

> sorry, jumped the gun here. didn't catch that this is still on hold for https://github.com/Expensify/App/issues/35011

No worries! I honestly have no idea if this behavior is still reproducible. I think we should pause on getting new proposals and instead get a retest to see if the behavior persists.

**Comment by mukhrr** [comment]
## Proposal

### What is the root cause of that problem?

The green "New messages" pill only appears when the unread marker sits below the current viewport (the user has scrolled up past it):

https://github.com/Expensify/App/blob/4bafaca569810b9910367608e49d6333c135e73a/src/pages/inbox/report/useReportUnreadMessageScrollTracking.ts#L148-L151

but clicking it runs `scrollToBottomAndMarkReportAsRead`, which unconditionally scrolls to the newest message and never uses the unread marker position that the list already knows (`unreadMarkerReportActionID` is right there in the same hook):

https://github.com/Expensify/App/blob/4bafaca569810b9910367608e49d6333c135e73a/src/hooks/useReportActionsScroll.ts#L343-L356

https://github.com/Expensify/App/blob/4bafaca569810b9910367608e49d6333c135e73a/src/pages/inbox/report/ReportActionsList.tsx#L448-L451

I verified this with the bug live, watching the marker's on-screen position through the click:

```
mark a message as unread (~2 screens above bottom)  → "New" line renders above it     marker rect top = 331 (in viewport)
scroll to the top of the conversation               → marker rect top = 2131          below viewport → green "New messages" pill appears
click the pill                                      → scrollToBottom() runs           useReportActionsScroll.ts L354
after the jump                                      → marker rect top = −742          above viewport; newest message on screen
```

The click flies past the entire unread range and also marks the report read, so the user is left at the bottom with no cue where the new messages started. It only *looks* correct when the marked message happens to sit within the last screen of the chat (landing at the bottom incidentally reveals the marker) — which is why the first click can appear to work while a later click, from further away, visibly fails.

### What changes do you think we should make in order to solve the problem?

In `scrollToBottomAndMarkReportAsRead` (`src/hooks/useReportActionsScroll.ts`), when an unread marker exists, scroll to the marker's index instead of the bottom, using the same positioning the app already uses for linked-message and action-badge jumps (`viewPosition: 1` with `CONST.REPORT.ACTIONS.LINKED_MESSAGE_OFFSET`). Resolve the index against `renderedVisibleReportActions` via `keyExtractor` (the FlatList data domain, same as the mount-scroll logic) rather than trusting `unreadMarkerReportActionIndex`, since a synthetic Concierge draft can shift rendered indexes. Fall back to `scrollToBottom()` when there is no marker or the marker is the newest action, and keep calling `markNewestActionAsRead()` — it clears the unread state without resetting the session's marker line (`readNewestAction` only resets it when `shouldResetUnreadMarker` is passed), so the "New" divider stays visible while the user reads.

```diff
-        reportScrollManager.scrollToBottom();
+        const unreadMarkerIndex = unreadMarkerReportActionID ? renderedVisibleReportActions.findIndex((item) => keyExtractor(item) === unreadMarkerReportActionID) : -1;
+        if (unreadMarkerIndex > 0) {
+            reportScrollManager.scrollToIndex(unreadMarkerIndex, {viewPosition: 1, viewOffset: CONST.REPORT.ACTIONS.LINKED_MESSAGE_OFFSET});
+        } else {
+            reportScrollManager.scrollToBottom();
+        }
         markNewestActionAsRead();
```

With this change the identical browser flow lands the viewport exactly at the marker (observed marker rect top = 103 after the click, "New" line visible above the first unread message), and it works on every subsequent click too since the marker persists for the session. I also added a regression test to the existing `tests/unit/useReportActionsScrollTest.tsx` harness asserting the pill click calls `scrollToIndex` with the marker index instead of `scrollToBottom` — it fails on current main and passes with the fix.

### What alternative solutions did you explore? (Optional)

- Skipping `markNewestActionAsRead()` when jumping to the marker so the report stays unread until the user reaches the bottom — rejected: it changes LHN unread behavior beyond what this issue asks, and the marker line already survives the read call, so nothing is gained.
- Using the provided `unreadMarkerReportActionIndex` directly — works in the common case but is computed over `sortedVisibleReportActions`, which can drift one index from the rendered data when a synthetic Concierge draft is inserted, so resolving against the rendered list is safer.
- `MoneyRequestReportActionsList` has its own copy of this pill handler with the same scroll-to-bottom behavior; I scoped the fix to chat reports per the issue, but the same change can be applied there if the team wants it.

https://github.com/user-attachments/assets/5570e507-77ed-4a83-bdda-0ece9258f979

https://github.com/user-attachments/assets/b300a817-fb08-49d0-9e18-680c1ab88114

**Comment by mukhrr** [comment]
@mvtglobally  I can still reproduce on latest `main`
@Ollyws mind reviewing [proposal](https://github.com/Expensify/App/issues/12249#issuecomment-5314678694), pls?! thanks

**Comment by mallenexpensify** [comment]
@Ollyws what's status here?  I feel like we should _do something_ to move it forward or close it.

**Comment by Ollyws** [comment]
Well looking at the code it seems that we are explicitly scrolling to the bottom rather than to the unread message, but if the expected behaviour is to scroll to the actual unread marker then @emkhalid's [proposal](https://github.com/Expensify/App/issues/12249#issuecomment-4970685866) LGTM.

🎀👀🎀 C+ reviewed

**Comment by emkhalid** [comment]
@lakchote we are waiting for your review and assignment based on [this](https://github.com/Expensify/App/issues/12249#issuecomment-5491570201), Thanks!

**Comment by lakchote** [comment]
@emkhalid's proposal LGTM

**Comment by emkhalid** [comment]
Thanks I will raise the PR.

**Comment by emkhalid** [comment]
@Ollyws the PR is ready for your review, Thanks!

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Fifth week)
